.evergreen/vs-env-run.ps1

```diff
@@ -0,0 +1,247 @@
+$vsdevcmd_bat = @(Get-ChildItem `
+        -Path $selected.installationPath `
+        -Filter "VsDevCmd.bat" -Recurse)[0]
+
+$env_script_content = ""
+
+# Use batch and the 'set' command to get the required environment variables out.
+if ($null -eq $vsdevcmd_bat) {
+    Write-Warning "No VsDevCmd.bat found for the requested VS version. Falling back to vcvarsall.bat"
+    Write-Warning "Additional platform selection functionality will be limited"
+    $vcvarsall_bat = @(Get-ChildItem -Path $selected.installationPath -Filter "vcvarsall.bat" -Recurse)[0]
+    if ($null -eq $vcvarsall_bat) {
+        throw "No VsDevCmd.bat nor vcvarsall.bat file found for requested Visual Studio version '$($selected.installationVersion)'"
+    }
+    Write-Debug "Using vcvarsall: [$($vcvarsall_bat.FullName)]"
+    $env_script_content = @"
+        @echo off
+        call "$($vcvarsall_bat.FullName)" $TargetArch $WinSDKVersion
+        set _rc=%ERRORLEVEL%
+        set
+        exit /b %_rc%
+"@
+}
+else {
+    # Build up the argument string to load the appropriate environment
+    $argstr = "-no_logo -arch=$TargetArch -host_arch=$HostArch -app_platform=$AppPlatform"
+    if ($UseSpectreMiticationLibraries) {
+        $argstr += " -vcvars_specter_libs=spectre"
+    }
+    if ($WinSDKVersion) {
+        $argstr += " -winsdk=$WinSDKVersion"
+    }
+    if ($VCToolsetVersion) {
+        $argstr += " -vcvars_ver=$VCToolsetVersion"
+    }
+
+    Write-Debug "Using VsDevCmd: [${vsdevcmd_bat.FullName}]"
+    $env_script_content = @"
+        @echo off
+        call "$($vsdevcmd_bat.FullName)" $argstr
+        set _rc=%ERRORLEVEL%
+        set
+        exit /b %_rc%
+"@
+}
+
+# Write the script and then execute it, capturing its output
+Set-Content "$ScratchDir/.env.bat" $env_script_content
+
+Write-Host "Loading VS environment..."
+$output = & cmd.exe /c "$ScratchDir/.env.bat"
+if ($LASTEXITCODE -ne 0) {
+    throw "Loading the environment failed [$LASTEXITCODE]:`n$output"
+}
+
+# The plain 'set' command emits VAR=VALUE lines for each loaded environment
+# variable. Parse those out and set them in our own environment.
+$prior_env = @{}
+foreach ($line in $output.Split("`r`n")) {
+    if ($line -match "(\w+)=(.+)") {
+        $varname = $Matches[1]
+        $value = $Matches[2]
+        # Set the environment value (may be null):
+        $prior_env[$varname] = [System.Environment]::GetEnvironmentVariable($varname)
+        [System.Environment]::SetEnvironmentVariable($varname, $value)
+    }
+}
+
+Write-Debug "Running command: $(ConvertTo-Json $Command)"
+
+try {
+    # Now invoke the external command.
+    & $Command
+    if ($LASTEXITCODE -ne 0) {
+        # If it was a process that returned non-zero, throw an error.
+        throw "Subcommand failed [$LASTEXITCODE]"
+    }
+}
+finally {
+    # Restore the prior environment
+    foreach ($key in $prior_env.Keys) {
+        [System.Environment]::SetEnvironmentVariable($key, $prior_env[$key])
```

Comment:
Given this script is intended to be invoked as a subshell, is it really necessary to include logic to preserve and restore environment variables in this manner?

Reply:
In the case of a subprocess, no, but if you invoke this script from an existing PS script/repl (without spawning a subprocess), it will share the environment variables, so you have to put them back before returning.

Reply:
No scripts in this PR use this script in such a manner, but I use it frequently in my own workflow both from REPL and other PS scripts.